ipwb/replay.py

```diff
@@ -179,6 +180,107 @@ def showMementosForURIRs_sansJS():
     return redirect('/memento/*/' + urir, code=301)
 
 
+def bin_search(iter, key, datetime=None):
+    # Read line encompassing current position
+    ln = iter.readline()
+
+    while ipwbUtils.isCDXJMetadataRecord(ln):
+        ln = iter.readline()
+
+    surtk, datetimeK, rest = ln.split(maxsplit=2)
+
+    matchDegree = getMatchDegree(surt, datetime, surtk, datetimeK)
+    if matchDegree == MementoMatch.EXACTMATCH:
+        return [ln]
+
+    # If further searching required...
+    left = 0
+    # Go to end of seek stream
+    iter.seek(0, 2)
+    right = iter.tell()  # Current position
+
+    lines = set()  # Prevents dupes
+    while (right - left > 1):
+        mid = (right + left) // 2
+        iter.seek(mid)
+        iter.readline()  # Purge rest of current line
+        line = iter.readline()  # Read the next full line
+
+        if len(line) == 0:
+            right = mid
+            continue
+
+        surtk, datetimeK, rest = line.split(maxsplit=2)
+
+        surtk = surtk.rstrip(b"/")
+        key = key.rstrip(b"/")
```

Comment:
Move this `key` sanitization outside of the while loop so it is done once as it is a received function argument, not something that is generated on each line like `surtk`.

Reply:
@ibnesayeed Outside of the preceding while loop?

https://github.com/oduwsdl/ipwb/blob/8cb614408d740670a923abc992169fb076630df5/ipwb/replay.py#L203-L216

This snippet looks like that cannot be done due to the dependency on the parameters of the while loop established just before finishing each iteration and reuse on the next.

Reply:
It is not clear to me why we cannot move `key = key.rstrip(b"/")` line before the outer loop starts. In fact we should be able to place it immediately after function body begins.

Reply:
Ah, ok, you are correct, @ibnesayeed . `key` sanitization can be moved before the outer loop. I misread that originally as both `key` and `surtk` sanitization, the latter of which depends on operations within the `while`.

Reply:
I thought I made the difference in `key` and `surtk` clear in the first comment. :)

Reply:
Applied in fa1b68c.

Reply:
@ibnesayeed You did. I misinterpreted `key` for key where the latter might refer to both `surtk` and `key`.